### PR TITLE
Update sinon to use shipped fake-timers types

### PR DESCRIPTION
**Pull request by 43081j**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/sinonjs/fake-timers#help-us-get-our-typescript-definitions-production-ready
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

---

I have:

* Added `@sinonjs/fake-timers` to the dependency whitelist - microsoft/DefinitelyTyped-tools#233
* Pushed a branch to sinon which fixes their JSDoc types - sinonjs/fake-timers#370 (**MUST BE MERGED BEFORE THIS CAN BUILD**)
* Bumped the major version
* Added package.json to depend on `@sinonjs/fake-timers` (the only DT package using it is sinon)

Will also leave some comments where i want feedback.

Draft for now.

**Sinon's own types are a lot weaker than the ones we previously had**, so this will actually introduce _worse_ types but at least they will work. Meanwhile I will contribute up to sinon some better JSDoc types if i can find time, to align with what we used to have.

cc @sandersn hopefully this is what you had in mind in #52294
cc @bcoe

### CODE REVIEW

File: types/sinon/index.d.ts
Comment (43081j) [inline types/sinon/index.d.ts:778]: this doesn't exist anymore. right now sinon's types define the return type as `any`, but i imagine ultimately it will instead just be `number|NodeTimer` (whatever the interface of node's timers is called)

File: types/sinon/index.d.ts
Comment (43081j) [inline types/sinon/index.d.ts:1660]: this correctly will NOT have a `restore()` method - clocks made via `clock.create()` are "raw" fake-timers clocks which have no `restore` method (but do have `uninstall`).

clocks made via `useFakeTimers` in sinon have `restore` set to `clock.uninstall` (as an alias).

File: types/sinon/index.d.ts
Comment (sandersn) [inline types/sinon/index.d.ts:14]: shouldn't this need a package.json now to install the real `@sinonjs/fake-timers`?

File: types/sinon/index.d.ts
Comment (sandersn) [inline types/sinon/index.d.ts:1]: It's very likely the right thing. But bumping the major version also implies that the rest of these types are now 10.0. Is that the case? If it's not then this will need a followup PR pretty quickly to make it so.

**Review by sandersn — CHANGES_REQUESTED** [review]
Seems like it's on the right track, but I think package.json is missing, and there may be required followup work to get the rest of sinon's types to 10.0.

File: types/sinon/index.d.ts
Comment (43081j) [reply types/sinon/index.d.ts:14]: i think i must've just forgot to check it in, i do have one locally :D ill sort it out 👍

File: types/sinon/index.d.ts
Comment (43081j) [reply types/sinon/index.d.ts:1]: it _should_ be, but i will double check. i think the versions were already diverged and 9 of the types was actually 10 of the code. ill make sure though and get back to you

**Comment by bcoe** [comment]
@sandersn @43081j I've tested this fix in these three repos, which use sinon (Including the repo _nodejs-storage_ which uses timers):

* https://github.com/googleapis/release-please/blob/master/test/release-pr.ts
* https://github.com/googleapis/nodejs-secret-manager/blob/master/test/gapic_secret_manager_service_v1.ts
* https://github.com/googleapis/nodejs-storage/blob/efa5bb8a22ab68c0c3e8549850e5db4d57ff29bb/test/signer.ts

I can't guarantee this won't break some folks, but the patch in our case gets all the repos compiling again.

**Comment by sandersn** [comment]
@43081j Crazy idea, how hard would it be to ship a fix to the sinon types that works with the broken types in fake-timer and the fix you proposed in  sinonjs/fake-timers#370? That way we could ship this fix and it'll pick up the corrected fake-timers types when it ships a new version.

**Comment by 43081j** [comment]
thats what i was thinking @sandersn, if i understood correctly...

* ship this with the weaker types (depending on fake-timers' own types)
* and the PR i submitted to fake-timers so we at least have the _right_ types (they are wrong at the min)

then later on PR the rest of the type improvements to fake-timers own types which would hopefully be automatically picked up by DT?

**Comment by 43081j** [comment]
In case either of you are interested, i've also opened sinonjs/fake-timers#371 to try fix the remaining types.

**Comment by sandersn** [comment]
Could you make changes so that it even compiles with fake-timers' currently-incorrect types? Otherwise this PR is blocked on sinonjs/fake-timers#370.

**Comment by 43081j** [comment]
I could but one or more of the functions have their parameters the wrong way around, so we would be publishing pretty broken types. 

If I comment the tests out though it'll probably build that way.

**Comment by 43081j** [comment]
FYI between v9 and v10 he seems to have pretty much re-formatted the sources and changed to newer syntax (template literals etc). So im fairly sure the types are already correct with the version bump

**Comment by 43081j** [comment]
@sandersn a new version of fake-timers has been pushed with my PR merged.

however, until/if my other PR gets merged (sinonjs/fake-timers#371) the following 3 tests fail here:

```ts
    now = clock.setTimeout(fnWithArgs, 0, 1234, 'abc');
    now = clock.setInterval(fnWithArgs, 0, 1234, 'abc');
    now = clock.setImmediate(fnWithArgs, 1234, 'abc');
```

because the types are still wrong in that they don't allow arbitrary arguments.

we could comment these tests out to get a working set of types published, knowing these 3 functions wont accept extra args for now. or we could wait until my other PR merges.

what do you think?

**Comment by sandersn** [comment]
I would comment out the tests and then send a followup that re-enables them after sinonjs/fake-timers#371 is merged.

**Review by sandersn — APPROVED** [review]

**Comment by 43081j** [comment]
Ready to merge

🥳

**Pull request merged**
